ramips: add support for Blueendless Kimax U35WF

arvati opened the pull request:
Blueendless Kimax U35WF is a 3,5" HDD Enclosure with Wi-Fi and Ethernet

Patch rewritten from: https://forum.openwrt.org/viewtopic.php?id=66908
Based on: https://github.com/lede-project/source/pull/965

Specification:
- SoC: MediaTek MT7620N
- CPU/Speed: 580 MHz
- Flash-Chip: KH25L12835F Spi Flash
- Flash size: 16 MiB
- RAM: 64 MiB
- LAN: 1x 100 Mbps Ethernet
- WiFi SoC-integrated: 802.11bgn
- 1x USB 2.0
- UART: for serial console

Installation:
1. Download sysupgrade.bin
2. Open vendor web interface
3. Choose to upgrade firmware
3. After reboot connect via ethernet at 192.168.1.1

Signed-off-by: Ademar Arvati Filho <arvati@hotmail.com>

arvati committed:
ramips: add support for Blueendless Kimax U35WF

Blueendless Kimax U35WF is a 3,5" HDD Enclosure with Wi-Fi and Ethernet

Patch rewritten from: https://forum.openwrt.org/viewtopic.php?id=66908
Based on: https://github.com/lede-project/source/pull/965

Specification:
- SoC: MediaTek MT7620N
- CPU/Speed: 580 MHz
- Flash-Chip: KH25L12835F Spi Flash
- Flash size: 16 MiB
- RAM: 64 MiB
- LAN: 1x 100 Mbps Ethernet
- WiFi SoC-integrated: 802.11bgn
- 1x USB 2.0
- UART: for serial console

Installation:
1. Download sysupgrade.bin
2. Open vendor web interface
3. Choose to upgrade firmware
3. After reboot connect via ethernet at 192.168.1.1

Signed-off-by: Ademar Arvati Filho <arvati@hotmail.com>

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -244,6 +244,15 @@ define Device/u25awf-h1
 endef
 TARGET_DEVICES += u25awf-h1
 
+define Device/u35wf
```
Have you actually tested a sysupgrade? It doesn't work. Please use `define Device/kimax_u35wf` or `define Device/kimax_u35awf` depending on what you intended to do with the compatible string.

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -244,6 +244,15 @@ define Device/u25awf-h1
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-mt76 kmod-usb-ohci \
+  			kmod-usb-ledtrig-usbport kmod-scsi-core kmod-usb-storage
+endef
+TARGET_DEVICES += u35wf
```
Please use `TARGET_DEVICES += kimax_u35wf` or `TARGET_DEVICES += kimax_u35awf` depending on what you intended to do with the compatible string.

arvati replied:
my mistake, is supposed to be kimax_u35wf

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -244,6 +244,15 @@ define Device/u25awf-h1
+  DTS := U35WF
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
+  DEVICE_TITLE := Kimax U35WF
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-mt76 kmod-usb-ohci \
```
no usb LED, no need for `kmod-usb-ledtrig-usbport`. I'm not sure about the other packages. Usually we only add `kmod-usb2 kmod-usb-ohci` for usb. But considering the purpose of the board the other packages kind of make sense.

arvati replied:
reviewed, thanks

mkresin commented on target/linux/ramips/dts/U35WF.dts:
```diff
@@ -0,0 +1,106 @@
+&pinctrl {
+	state_default: pinctrl0 {
+		default {
+			ralink,group = "ephy", "wled", "uartf";
```
No pin of the `uartf` group is used. Drop it. Have a look at https://kresin.me/files/pinmux_review_ramips.ods for reference.

arvati replied:
thanks again, i will drop it.

mkresin reviewed (changes requested):
Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

arvati commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -263,6 +263,11 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"4:lan" "6t@eth0"
 		;;
+	kimax,u35wf)
```
thanks, change made it.

arvati commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -20,7 +20,8 @@ alfa-network,awusfree1)
 all0256n-8M|\
-all5002)
+all5002|\
+kimax,u35wf)
```
ok, it was just to make it readable, change reverted.

arvati commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -244,6 +244,15 @@ define Device/u25awf-h1
 
+define Device/u35wf
+  DTS := U35WF
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
```
thanks again

arvati commented on target/linux/ramips/dts/U35WF.dts:
```diff
@@ -0,0 +1,106 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "kimax,u35awf","ralink,mt7620n-soc";
```
my mistake is supposed to be kimax,u35wf

arvati commented:
Thanks for reviewing mkresin, i just updated commit with changes.

obsy commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -244,6 +244,15 @@ define Device/u25awf-h1
+  IMAGE_SIZE := 16064k
+  DEVICE_TITLE := Kimax U35WF
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-ohci \
+  			kmod-scsi-core kmod-usb-storage
```
Should be filesystem added as well? vfat/ext4?

arvati replied:
I will make one more change, excluding these packages since they are not needed for booting. I will commit back once tested again later. I am not with device right now, just in couple hours.

arvati commented:
changes commited, see dmesg result with plain config (and enabled wifi after first login) bellow:
```
dmesg
[    0.000000] Linux version 4.14.52 (root@openwrt) (gcc version 7.3.0 (OpenWrt GCC 7.3.0 r7383-ee6680c245)) #0 Wed Jul 4 11:28:53 2018
[    0.000000] Board has DDR1
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7620N ver:2 eco:6
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
[    0.000000] MIPS: machine is Kimax U35WF
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 04000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000] free_area_init_node: node 0, pgdat 8047cf80, node_mem_map 81000040
[    0.000000]   Normal zone: 128 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] random: get_random_bytes called from start_kernel+0x8c/0x47c with crng_init=0
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16256
[    0.000000] Kernel command line: console=ttyS0,57600 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 256 (order: -2, 1024 bytes)
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Writing ErrCtl register=0001ec21
[    0.000000] Readback ErrCtl register=0001ec21
[    0.000000] Memory: 59860K/65536K available (3561K kernel code, 183K rwdata, 856K rodata, 192K init, 222K bss, 5676K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] CPU Clock: 580MHz
[    0.000000] clocksource: systick: mask: 0xffff max_cycles: 0xffff, max_idle_ns: 583261500 ns
[    0.000000] systick: enable autosleep mode
[    0.000000] systick: running - mult: 214748, shift: 32
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000011] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.015468] Calibrating delay loop... 385.84 BogoMIPS (lpj=1929216)
[    0.087717] pid_max: default: 32768 minimum: 301
[    0.097130] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.110154] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.128851] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.148323] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.160509] pinctrl core: initialized pinctrl subsystem
[    0.172017] NET: Registered protocol family 16
[    0.204017] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.215039] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.227597] rt2880_gpio 10000660.gpio: registering 32 gpios
[    0.238586] rt2880_gpio 10000660.gpio: registering 32 irq handlers
[    0.251177] rt2880_gpio 10000688.gpio: registering 1 gpios
[    0.261990] rt2880_gpio 10000688.gpio: registering 1 irq handlers
[    0.278313] clocksource: Switched to clocksource systick
[    0.289950] NET: Registered protocol family 2
[    0.299526] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.313298] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.325886] TCP: Hash tables configured (established 1024 bind 1024)
[    0.338649] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.350152] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.362864] NET: Registered protocol family 1
[    0.371427] PCI: CLS 0 bytes, default 32
[    0.373685] rt-timer 10000100.timer: maximum frequency is 1220Hz
[    0.386350] Crashlog allocated RAM at address 0x3f00000
[    0.398001] workingset: timestamp_bits=30 max_order=14 bucket_order=0
[    0.416592] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.428085] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.455879] io scheduler noop registered
[    0.463574] io scheduler deadline registered (default)
[    0.474658] Serial: 8250/16550 driver, 16 ports, IRQ sharing enabled
[    0.490170] console [ttyS0] disabled
[    0.497162] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 20, base_baud = 2500000) is a Palmchip BK-3103
[    0.516925] console [ttyS0] enabled
[    0.530705] bootconsole [early0] disabled
[    0.547248] cacheinfo: Failed to find cpu0 device node
[    0.557518] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.575732] spi spi0.0: force spi mode3
[    0.584164] m25p80 spi0.0: mx25l12805d (16384 Kbytes)
[    0.594323] 4 fixed-partitions partitions found on MTD device spi0.0
[    0.606983] Creating 4 MTD partitions on "spi0.0":
[    0.616542] 0x000000000000-0x000000030000 : "u-boot"
[    0.627312] 0x000000030000-0x000000040000 : "u-boot-env"
[    0.638823] 0x000000040000-0x000000050000 : "factory"
[    0.649803] 0x000000050000-0x000001000000 : "firmware"
[    0.780171] 2 uimage-fw partitions found on MTD device firmware
[    0.792006] 0x000000050000-0x0000001c8993 : "kernel"
[    0.802769] 0x0000001c8993-0x000001000000 : "rootfs"
[    0.813487] mtd: device 5 (rootfs) set to be root filesystem
[    0.826080] 1 squashfs-split partitions found on MTD device rootfs
[    0.838468] 0x0000003c0000-0x000001000000 : "rootfs_data"
[    0.850835] libphy: Fixed MDIO Bus: probed
[    0.863235] gsw: setting port4 to ephy mode
[    0.871638] mtk_soc_eth 10100000.ethernet eth0 (uninitialized): port 4 link up (100Mbps/Full duplex)
[    0.890047] mtk_soc_eth 10100000.ethernet: loaded mt7620 driver
[    0.902527] mtk_soc_eth 10100000.ethernet eth0: mediatek frame engine at 0xb0100000, irq 5
[    0.919512] rt2880_wdt 10000120.watchdog: Initialized
[    0.931006] NET: Registered protocol family 10
[    0.943476] Segment Routing with IPv6
[    0.950955] NET: Registered protocol family 17
[    0.959869] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[    0.985692] 8021q: 802.1Q VLAN Support v1.8
[    1.008622] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    1.023702] Freeing unused kernel memory: 192K
[    1.032574] This architecture does not have kernel memory protection.
[    2.405074] init: Console is alive
[    2.412146] init: - watchdog -
[    2.543386] random: fast init done
[    4.235335] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    4.500293] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    4.518786] init: - preinit -
[    6.173478] 8021q: adding VLAN 0 to HW filter on device eth0
[    6.331487] random: procd: uninitialized urandom read (4 bytes read)
[    7.810076] jffs2: notice: (358) jffs2_build_xattr_subsystem: complete building xattr subsystem, 1 of xdatum (0 unchecked, 1 orphan) and 1 of xref (1 dead, 0 orphan) found.
[    7.843587] mount_root: switching to jffs2 overlay
[    7.881100] overlayfs: upper fs does not support tmpfile.
[    7.901865] urandom-seed: Seeding with /etc/urandom.seed
[    8.165350] procd: - early -
[    8.171977] procd: - watchdog -
[    8.502150] procd: - watchdog -
[    8.508805] procd: - ubus -
[    8.675763] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.695286] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.708507] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.722132] procd: - init -
[    9.262047] kmodloader: loading kernel modules from /etc/modules.d/*
[    9.281038] ip6_tables: (C) 2000-2006 Netfilter Core Team
[    9.304480] Loading modules backported from Linux version wt-2017-11-01-0-gfe248fc2c180
[    9.320498] Backport generated by backports.git v4.14-rc2-1-31-g86cf0e5d
[    9.338164] ip_tables: (C) 2000-2006 Netfilter Core Team
[    9.358188] nf_conntrack version 0.5.0 (1024 buckets, 4096 max)
[    9.436307] xt_time: kernel timezone is -0000
[    9.512554] PPP generic driver version 2.4.2
[    9.523862] NET: Registered protocol family 24
[    9.563315] rt2800_wmac 10180000.wmac: loaded eeprom from mtd device "factory"
[    9.577767] ieee80211 phy0: rt2x00_set_rt: Info - RT chipset 6352, rev 0500 detected
[    9.593226] ieee80211 phy0: rt2x00_set_rf: Info - RF chipset 7620 detected
[    9.607593] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[    9.682632] kmodloader: done loading kernel modules from /etc/modules.d/*
[   10.876080] urandom_read: 5 callbacks suppressed
[   10.876089] random: jshn: uninitialized urandom read (4 bytes read)
[   18.364391] 8021q: adding VLAN 0 to HW filter on device eth0
[   18.387957] br-lan: port 1(eth0) entered blocking state
[   18.398438] br-lan: port 1(eth0) entered disabled state
[   18.409200] device eth0 entered promiscuous mode
[   18.473693] br-lan: port 1(eth0) entered blocking state
[   18.484162] br-lan: port 1(eth0) entered forwarding state
[   18.495118] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   19.422917] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   21.638855] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[   21.656834] br-lan: port 2(wlan0) entered blocking state
[   21.667552] br-lan: port 2(wlan0) entered disabled state
[   21.678516] device wlan0 entered promiscuous mode
[   21.698458] IPv6: ADDRCONF(NETDEV_CHANGE): wlan0: link becomes ready
[   21.711349] br-lan: port 2(wlan0) entered blocking state
[   21.721968] br-lan: port 2(wlan0) entered forwarding state
[   45.984971] random: crng init done
```

mkresin commented:
merged, thanks!

alazyworkaholic commented:
Thanks Avarti, mkresin! Should we be able to use the final 18.06.0 release with the U35WF, or will we have to wait for the one after that?
